ppl/archive/import.go

```diff
@@ -53,34 +56,53 @@ func Import(ctx context.Context, ark *Archive, zctx *resolver.Context, r zbuf.Re
 // provided read stream is closed. Add some kind of timeout functionality that
 // periodically flushes stale tsDirWriters.
 type Writer struct {
-	ark     *Archive
-	ctx     context.Context
-	writers map[tsDir]*tsDirWriter
+	ark      *Archive
+	cancel   context.CancelFunc
+	ctx      context.Context
+	errgroup *errgroup.Group
+	mu       sync.RWMutex
+	once     sync.Once
+	writers  map[tsDir]*tsDirWriter
 
 	memBuffered int64
 	stats       ImportStats
 }
 
 func NewWriter(ctx context.Context, ark *Archive) *Writer {
+	g, ctx := errgroup.WithContext(ctx)
+	ctx, cancel := context.WithCancel(ctx)
 	return &Writer{
-		ark:     ark,
-		ctx:     ctx,
-		writers: make(map[tsDir]*tsDirWriter),
+		ark:      ark,
+		cancel:   cancel,
+		ctx:      ctx,
+		errgroup: g,
+		writers:  make(map[tsDir]*tsDirWriter),
 	}
 }
 
 func (w *Writer) Write(rec *zng.Record) error {
+	w.once.Do(func() {
+		w.errgroup.Go(w.flusher)
+	})
+	if w.ctx.Err() != nil {
+		return w.errgroup.Wait()
```

Comment:
Why not `return w.ctx.Err()`?

Reply:
I wanted to propogate the errgroup err if the context cancellation is called by the errgroup. That said, I should probably return ctx.Err() if errgroup.Wait() is nil.